Crystal interpreter cannot find pthread

## Bug Report

Trying to run `crystal i` fails with the following exception:
<details>
<summary>Exception</summary>

```
cannot find -lpthread
Linker arguments: -L/usr/lib64/crystal -lpthread -levent -lrt -ldl
Search path: /usr/lib64/llvm12/lib:/lib64:/usr/lib64:/lib:/usr/lib:/usr/lib64/crystal (Crystal::Loader::LoadError)
  from src/compiler/crystal/loader.cr:74:31 in 'load_library'
  from src/compiler/crystal/loader.cr:42:7 in 'new'
  from src/compiler/crystal/loader/unix.cr:51:7 in 'parse'
  from src/compiler/crystal/interpreter/context.cr:365:1 in 'loader'
  from src/compiler/crystal/interpreter/context.cr:378:5 in 'c_function'
  from src/compiler/crystal/interpreter/compiler.cr:1860:17 in 'compile_lib_call'
  from src/compiler/crystal/interpreter/compiler.cr:1764:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:2977:5 in 'accept_with_wants_value'
  from src/compiler/crystal/interpreter/compiler.cr:2967:5 in 'request_value'
  from src/compiler/crystal/interpreter/compiler.cr:578:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:2977:5 in 'accept_with_wants_value'
  from src/compiler/crystal/interpreter/compiler.cr:2967:5 in 'request_value'
  from src/compiler/crystal/interpreter/compiler.cr:1219:5 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:566:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:234:5 in 'compile_def'
  from src/compiler/crystal/interpreter/compiler.cr:1956:7 in 'create_compiled_def'
  from src/compiler/crystal/interpreter/compiler.cr:1777:22 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:234:5 in 'compile_def'
  from src/compiler/crystal/interpreter/compiler.cr:1956:7 in 'create_compiled_def'
  from src/compiler/crystal/interpreter/compiler.cr:1777:22 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:447:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:234:5 in 'compile_def'
  from src/compiler/crystal/interpreter/compiler.cr:1399:5 in 'get_const_index_and_compiled_def'
  from src/compiler/crystal/interpreter/compiler.cr:2463:27 in 'initialize_const_if_needed'
  from src/compiler/crystal/interpreter/compiler.cr:1371:9 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:2977:5 in 'accept_with_wants_value'
  from src/compiler/crystal/interpreter/compiler.cr:2967:5 in 'request_value'
  from src/compiler/crystal/interpreter/compiler.cr:2121:9 in 'compile_call_args'
  from src/compiler/crystal/interpreter/compiler.cr:1773:5 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:234:5 in 'compile_def'
  from src/compiler/crystal/interpreter/compiler.cr:1399:5 in 'get_const_index_and_compiled_def'
  from src/compiler/crystal/interpreter/compiler.cr:680:31 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:2881:5 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:566:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:2977:5 in 'accept_with_wants_value'
  from src/compiler/crystal/interpreter/compiler.cr:2971:5 in 'discard_value'
  from src/compiler/crystal/interpreter/compiler.cr:2829:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:566:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:234:5 in 'compile_def:closure_owner'
  from src/compiler/crystal/interpreter/compiler.cr:2929:5 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:566:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:234:5 in 'compile_def:closure_owner'
  from src/compiler/crystal/interpreter/compiler.cr:2929:5 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:566:7 in 'visit'
  from src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from src/compiler/crystal/interpreter/compiler.cr:165:5 in 'compile'
  from src/compiler/crystal/interpreter/interpreter.cr:207:5 in 'interpret'
  from src/compiler/crystal/interpreter/repl.cr:146:5 in 'interpret'
  from src/compiler/crystal/interpreter/repl.cr:150:5 in 'interpret_and_exit_on_error'
  from src/compiler/crystal/interpreter/repl.cr:140:5 in 'load_prelude'
  from src/string.cr:278:18 in 'run'
  from src/compiler/crystal/command/repl.cr:10:7 in 'repl'
  from src/compiler/crystal/command.cr:102:7 in 'run'
  from src/compiler/crystal/command.cr:51:5 in 'run'
  from src/compiler/crystal/command.cr:50:3 in 'run'
  from src/compiler/crystal.cr:11:1 in '__crystal_main'
  from src/crystal/main.cr:115:5 in 'main_user_code'
  from src/crystal/main.cr:101:7 in 'main'
  from src/crystal/main.cr:127:3 in 'main'
  from /lib64/libc.so.6 in '??'
  from /lib64/libc.so.6 in '__libc_start_main'
  from /var/mnt/media/Projects/Other/crystal-git/.build/crystal in '_start'
  from ???
```
</details>

Before the new loader was merged, it worked. (1.3.2 works for me)

The pthread library is installed:
```
$ find /usr/lib -name "*pthread*"
/usr/lib/libpthread.so.0
$ find /usr/lib64 -name "*pthread*"
/usr/lib64/libpthread.so.0
/usr/lib64/pkgconfig/libevent_pthreads.pc
/usr/lib64/pkgconfig/pthread-stubs.pc
/usr/lib64/libevent_pthreads.so
/usr/lib64/libpthread.a
/usr/lib64/libpthread_nonshared.a
/usr/lib64/libevent_pthreads-2.1.so.7
/usr/lib64/libevent_pthreads-2.1.so.7.0.1

```

```
$ bin/crystal -v
Crystal 1.4.0-dev [197465c86] (2022-02-03)

LLVM: 13.0.0
Default target: x86_64-redhat-linux-gnu
```

Yes, that's all clear to me. Avoiding errors on a missing library makes things work when the library is already loaded in the compiler. But the error is intended. 

The problem with silencing these errors is when the library you want to link against is not already loaded in the compiler.

Take this example:
```cr
@[Link("foobar")]
lib LibFooBar
  fun foobar : Void
end

LibFooBar.foobar
```
If `libfoobar` does not exist, it errors with ``undefined reference to `foobar'``. This suggests `libfoobar` was correctly loaded but didn't provide the requested symbol. That's at least misleading, if not completely wrong. The error message should say that `libfoobar` could not be loaded: `cannot find -lfoobar`.

We should figure out *why* loading that library failed instead of adding another workaround on top of the existing workaround to even lookup symbols loaded in the compiler.

Could you look where your `libpthread.so` - that's supposed to be loaded - is located?

```
$ echo '               
@[Link("foobar")]
lib LibFooBar
  fun foobar : Void
end

LibFooBar.foobar
' > test.cr

$ crystal run ./test.cr
/usr/bin/ld: cannot find -lfoobar (this usually means you need to install the development package for libfoobar)
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o /var/home/blobcodes/.cache/crystal/crystal-run-test.tmp  -rdynamic -L/usr/lib64/crystal -lfoobar -lpcre -lm -lgc -lpthread -levent  -lrt -ldl`

# Using a random library, renamed to libfoobar.so
$ crystal run ./test.cr --link-flags="-L/home/blobcodes"
/usr/bin/ld: _main.o: in function `__crystal_main':
/var/home/blobcodes/test.cr:7: undefined reference to `foobar'
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o /var/home/blobcodes/.cache/crystal/crystal-run-test.tmp -L/home/blobcodes -rdynamic -L/usr/lib64/crystal -lfoobar -lpcre -lm -lgc -lpthread -levent  -lrt -ldl`
```

Huh, I thought the error was going to say "undefined reference to ..." but it actually says the missing library. I probably misremember things.

I guess now I also think it makes sense to have `Crystal::Loader` work the way it works now. We just have to figure out how to find `pthread`, `iconv`, etc., at least on mac, for interpreted mode. I assume it works on unix...

--- a/src/gc/boehm.cr
+++ b/src/gc/boehm.cr
@@ -5,9 +5,15 @@
 # MUSL: On musl systems, libpthread is empty. The entire library is already included in libc.
 # The empty library is only available for POSIX compatibility. We don't need to link it.
 #
+# Interpreter: Starting with glibc 2.34, `pthread` is integrated into `libc`
+# and may not even be available as a separate shared library.
+# There's always a static library for compiled mode, but `Crystal::Loader` does not support
+# static libraries. So we just skip `pthread` entirely. The symbols are still
+# available in the interpreter because they are loaded in the compiler.
+#
 # OTHERS: On other systems, we add the linker annotation here to make sure libpthread is loaded
 # before libgc which looks up symbols from libpthread.
-{% unless flag?(:win32) || flag?(:musl) %}
+{% unless flag?(:win32) || flag?(:musl) || (flag?(:interpreter) && flag?(:gnu)) %}
   @[Link("pthread")]
 {% end %}
 

--- a/src/gc/boehm.cr
+++ b/src/gc/boehm.cr
@@ -13,7 +13,7 @@
 #
 # OTHERS: On other systems, we add the linker annotation here to make sure libpthread is loaded
 # before libgc which looks up symbols from libpthread.
-{% unless flag?(:win32) || flag?(:musl) || (flag?(:interpreter) && flag?(:gnu)) %}
+{% unless flag?(:win32) || flag?(:musl) || (flag?(:interpreted) && flag?(:gnu)) %}
   @[Link("pthread")]
 {% end %}
 

--- a/src/gc/boehm.cr
+++ b/src/gc/boehm.cr
@@ -5,6 +5,8 @@
 # MUSL: On musl systems, libpthread is empty. The entire library is already included in libc.
 # The empty library is only available for POSIX compatibility. We don't need to link it.
 #
+# Darwin: `libpthread` is provided as part of `libsystem`. There's no reason to link it explicitly.
+#
 # Interpreter: Starting with glibc 2.34, `pthread` is integrated into `libc`
 # and may not even be available as a separate shared library.
 # There's always a static library for compiled mode, but `Crystal::Loader` does not support
@@ -13,7 +15,7 @@
 #
 # OTHERS: On other systems, we add the linker annotation here to make sure libpthread is loaded
 # before libgc which looks up symbols from libpthread.
-{% unless flag?(:win32) || flag?(:musl) || (flag?(:interpreted) && flag?(:gnu)) %}
+{% unless flag?(:win32) || flag?(:musl) || flag?(:darwin) || (flag?(:interpreted) && flag?(:gnu)) %}
   @[Link("pthread")]
 {% end %}
 

--- a/src/lib_c/aarch64-darwin/c/dlfcn.cr
+++ b/src/lib_c/aarch64-darwin/c/dlfcn.cr
@@ -1,4 +1,3 @@
-@[Link("dl")]
 lib LibC
   RTLD_LAZY    = 0x1
   RTLD_NOW     = 0x2

--- a/src/lib_c/x86_64-darwin/c/dlfcn.cr
+++ b/src/lib_c/x86_64-darwin/c/dlfcn.cr
@@ -1,4 +1,3 @@
-@[Link("dl")]
 lib LibC
   RTLD_LAZY    = 0x1
   RTLD_NOW     = 0x2

